src/afl-fuzz-init.c

```diff
@@ -589,8 +589,6 @@ void read_foreign_testcases(afl_state_t *afl, int first) {
         u8 *fn2 =
             alloc_printf("%s/%s", afl->foreign_syncs[iter].dir, nl[i]->d_name);
 
-        free(nl[i]);                                         /* not tracked */
```

Comment:
Couldn't you just have kept the pointer to `d_name` in a var here and kept the free here?
Or ar there other things that can happen?
Also, what about d_name, should it be freed or is it part of the struct or who owns it?

Reply:
you can not continue to use `nl[i]->d_name` once you have free'd `nl[i]`.
d_name seems not needed to be free'd - this is from the scandir man page:
```
       int
       main(void)
       {
           struct dirent **namelist;
           int n;

           n = scandir(".", &namelist, NULL, alphasort);
           if (n == -1) {
               perror("scandir");
               exit(EXIT_FAILURE);
           }

           while (n--) {
               printf("%s\n", namelist[n]->d_name);
               free(namelist[n]);
           }
           free(namelist);

           exit(EXIT_SUCCESS);
       }
```

Reply:
You're right, sorry